src/librustc/traits/error_reporting.rs

```diff
@@ -428,18 +428,20 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                     Some(format!("[{}]", self.tcx.type_of(def.did).to_string())),
                 ));
                 let tcx = self.tcx;
-                if let Some(len) = len.val.try_to_scalar().and_then(|scalar| {
-                    scalar.to_usize(&tcx).ok()
-                }) {
-                    flags.push((
-                        "_Self".to_owned(),
-                        Some(format!("[{}; {}]", self.tcx.type_of(def.did).to_string(), len)),
-                    ));
-                } else {
-                    flags.push((
-                        "_Self".to_owned(),
-                        Some(format!("[{}; _]", self.tcx.type_of(def.did).to_string())),
-                    ));
+                if let ty::LazyConst::Evaluated(len) = len {
+                    if let Some(len) = len.val.try_to_scalar().and_then(|scalar| {
+                        scalar.to_usize(&tcx).ok()
+                    }) {
+                        flags.push((
+                            "_Self".to_owned(),
+                            Some(format!("[{}; {}]", self.tcx.type_of(def.did).to_string(), len)),
+                        ));
+                    } else {
+                        flags.push((
+                            "_Self".to_owned(),
+                            Some(format!("[{}; _]", self.tcx.type_of(def.did).to_string())),
+                        ));
+                    }
```

Comment:
So we don't print anything when it is not evaluated? That is different from the previous code.

I have no idea where this code is called, just pointing out a functional change. ;)

Reply:
oops indeed. This is just for producing more diagnostics, but still. I found a nice way to fix it without more duplication.